primitives/src/v3/mod.rs

```diff
@@ -0,0 +1,1801 @@
+	/// in parachain consensus. See
+	/// [`max_validators`](https://github.com/paritytech/polkadot/blob/a52dca2be7840b23c19c153cf7e110b1e3e475f8/runtime/parachains/src/configuration.rs#L148).
+	///
+	/// `SessionInfo::validators` will be limited to to `max_validators` when set.
+	pub validators: IndexedVec<ValidatorIndex, ValidatorId>,
+	/// Validators' authority discovery keys for the session in canonical ordering.
+	///
+	/// NOTE: The first `validators.len()` entries will match the corresponding validators in
+	/// `validators`, afterwards any remaining authorities can be found. This is any authorities not
+	/// participating in parachain consensus - see
+	/// [`max_validators`](https://github.com/paritytech/polkadot/blob/a52dca2be7840b23c19c153cf7e110b1e3e475f8/runtime/parachains/src/configuration.rs#L148)
+	pub discovery_keys: Vec<AuthorityDiscoveryId>,
+	/// The assignment keys for validators.
+	///
+	/// NOTE: There might be more authorities in the current session, than validators participating
+	/// in parachain consensus. See
+	/// [`max_validators`](https://github.com/paritytech/polkadot/blob/a52dca2be7840b23c19c153cf7e110b1e3e475f8/runtime/parachains/src/configuration.rs#L148).
+	///
+	/// Therefore:
+	/// ```ignore
+	///		assignment_keys.len() == validators.len() && validators.len() <= discovery_keys.len()
+	///	```
+	pub assignment_keys: Vec<AssignmentId>,
+	/// Validators in shuffled ordering - these are the validator groups as produced
+	/// by the `Scheduler` module for the session and are typically referred to by
+	/// `GroupIndex`.
+	pub validator_groups: IndexedVec<GroupIndex, Vec<ValidatorIndex>>,
+	/// The number of availability cores used by the protocol during this session.
+	pub n_cores: u32,
+	/// The zeroth delay tranche width.
+	pub zeroth_delay_tranche_width: u32,
+	/// The number of samples we do of `relay_vrf_modulo`.
+	pub relay_vrf_modulo_samples: u32,
+	/// The number of delay tranches in total.
+	pub n_delay_tranches: u32,
+	/// How many slots (BABE / SASSAFRAS) must pass before an assignment is considered a
+	/// no-show.
+	pub no_show_slots: u32,
+	/// The number of validators needed to approve a block.
+	pub needed_approvals: u32,
+}
+
+// Structure upgrade
+impl From<v2::SessionInfo> for SessionInfo {
+	fn from(old: v2::SessionInfo) -> Self {
+		Self {
+			executor_params: ExecutorParams::default(),
+			active_validator_indices: old.active_validator_indices,
+			random_seed: old.random_seed,
+			dispute_period: old.dispute_period,
+			validators: old.validators,
+			discovery_keys: old.discovery_keys,
+			assignment_keys: old.assignment_keys,
+			validator_groups: old.validator_groups,
+			n_cores: old.n_cores,
+			zeroth_delay_tranche_width: old.zeroth_delay_tranche_width,
+			relay_vrf_modulo_samples: old.relay_vrf_modulo_samples,
+			n_delay_tranches: old.n_delay_tranches,
+			no_show_slots: old.no_show_slots,
+			needed_approvals: old.needed_approvals,
+		}
+	}
+}
```

Comment:
I would be in favor of putting this in the near of the migration and to not use `From`.

I would also actually start setting directly all the params to the current "configuration".

Reply:
The migration is not the only place where `From` implementation is used. In the "new node, old runtime" situation, we still need to be able to get an older structure from runtime and [upgrade it to a newer one](https://github.com/paritytech/polkadot/pull/6161/files#diff-1c816593e067f095a3388e13824a0729ac4d1d9b7fcf96ab2ea6761ba423a9dcR476).

Also, that is exactly [how it works for the v1->v2 transition](https://github.com/paritytech/polkadot/blob/88013730166ba90745ae7c9eb3e0c1be1513c7cc/primitives/src/v2/mod.rs#L1739).